### PR TITLE
Avoid incorrectly providing diagnostics for chained calls

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes https://github.com/aspnet/AspNetCore/issues/15203

### CODE REVIEW

**Comment by pranavkm** [comment]
@JamesNK, @centreboard started with a PR for this here: https://github.com/aspnet/AspNetCore/pull/15208/files. The issue is that Roslyn's callbacks for chained method calls appear in inverse order. Fixing the order is fairly difficult and the linked PR isn't complete.

This change avoids dealing with this issue in entirety by bailing out if we see an oddly ordered middleware.

**Comment by Tratcher** [comment]
That's unfortunate...
What's the estimated cost to actually fix this?

**Comment by pranavkm** [comment]
> What's the estimated cost to actually fix this?

It sort of follows from here: https://github.com/aspnet/AspNetCore/pull/15208/files#diff-41b957d6c50268c444338d72f9b7f41bR45, but you have to ensure that the current instance is a parent of the previous. The code's not impossibly hard to write, but since it's common across startup analyzers, I worry complicating it needs more bake time \ feedback time, which we don't have in preview3.

**Comment by pranavkm** [comment]
Sure, I'll file a follow up to fix this in 5.0.

File: src/Analyzers/Analyzers/src/UseAuthorizationAnalyzer.cs [hunk 1]
```diff
@@ -70,9 +72,23 @@ public void AnalyzeSymbol(SymbolAnalysisContext context)
...
                     {
+                        if (useEndpoint is null)
+                        {
+                            // We're likely here because the middleware uses an expression chain e.g.
+                            // app.UseRouting()
+                            //   .UseAuthorization()
+                            //   .UseEndpoints(..));
+                            // This analyzer expects MiddlewareItem instances to appear in the order in which they appear in source
+                            // which unfortunately isn't true for chained calls (the operations appear in reverse order).
+                            // We'll avoid doing any analysis in this event and rely on the runtime guardrails.
```
Comment (JamesNK) [inline src/Analyzers/Analyzers/src/UseAuthorizationAnalyzer.cs]: Create a 5.0 issue for handling this situation properly and link to it here

**Review by JamesNK — APPROVED** [review]

File: src/Analyzers/Analyzers/src/UseAuthorizationAnalyzer.cs [hunk 2]
```diff
@@ -70,9 +72,23 @@ public void AnalyzeSymbol(SymbolAnalysisContext context)
...
+                            return;
```
Comment (pranavkm) [inline src/Analyzers/Analyzers/src/UseAuthorizationAnalyzer.cs]:
```suggestion
                            // We'll use https://github.com/aspnet/AspNetCore/issues/16648 to track addressing this in a future milestone
                            return;
```

**Comment by Pilchie** [comment]
approved for 3.1.0-Preview3

**Pull request merged**